{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.83\n"
     ]
    }
   ],
   "source": [
    "import Bio\n",
    "print(Bio.__version__)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.83\n"
     ]
    }
   ],
   "source": [
    "print(Bio.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.SeqUtils in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.SeqUtils - Miscellaneous functions for dealing with sequences.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    CheckSum\n",
      "    IsoelectricPoint\n",
      "    MeltingTemp\n",
      "    ProtParam\n",
      "    ProtParamData\n",
      "    lcc\n",
      "\n",
      "CLASSES\n",
      "    builtins.dict(builtins.object)\n",
      "        CodonAdaptationIndex\n",
      "    \n",
      "    class CodonAdaptationIndex(builtins.dict)\n",
      "     |  CodonAdaptationIndex(sequences, table=NCBICodonTableDNA(id=1, names=['Standard', 'SGC0'], ...))\n",
      "     |  \n",
      "     |  A codon adaptation index (CAI) implementation.\n",
      "     |  \n",
      "     |  Implements the codon adaptation index (CAI) described by Sharp and\n",
      "     |  Li (Nucleic Acids Res. 1987 Feb 11;15(3):1281-95).\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      CodonAdaptationIndex\n",
      "     |      builtins.dict\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __init__(self, sequences, table=NCBICodonTableDNA(id=1, names=['Standard', 'SGC0'], ...))\n",
      "     |      Generate a codon adaptiveness table from the coding DNA sequences.\n",
      "     |      \n",
      "     |      This calculates the relative adaptiveness of each codon (w_ij) as\n",
      "     |      defined by Sharp & Li (Nucleic Acids Research 15(3): 1281-1295 (1987))\n",
      "     |      from the provided codon DNA sequences.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - sequences: An iterable over DNA sequences, which may be plain\n",
      "     |                    strings, Seq objects, MutableSeq objects, or SeqRecord\n",
      "     |                    objects.\n",
      "     |       - table:     A Bio.Data.CodonTable.CodonTable object defining the\n",
      "     |                    genetic code. By default, the standard genetic code is\n",
      "     |                    used.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  calculate(self, sequence)\n",
      "     |      Calculate and return the CAI (float) for the provided DNA sequence.\n",
      "     |  \n",
      "     |  optimize(self, sequence, seq_type='DNA', strict=True)\n",
      "     |      Return a new DNA sequence with preferred codons only.\n",
      "     |      \n",
      "     |      Uses the codon adaptiveness table defined by the CodonAdaptationIndex\n",
      "     |      object to generate DNA sequences with only preferred codons.\n",
      "     |      May be useful when designing DNA sequences for transgenic protein\n",
      "     |      expression or codon-optimized proteins like fluorophores.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |          - sequence: DNA, RNA, or protein sequence to codon-optimize.\n",
      "     |                      Supplied as a str, Seq, or SeqRecord object.\n",
      "     |          - seq_type: String specifying type of sequence provided.\n",
      "     |                      Options are \"DNA\", \"RNA\", and \"protein\". Default is \"DNA\".\n",
      "     |          - strict:   Determines whether an exception should be raised when\n",
      "     |                      two codons are equally prefered for a given amino acid.\n",
      "     |      Returns:\n",
      "     |          Seq object with DNA encoding the same protein as the sequence argument,\n",
      "     |          but using only preferred codons as defined by the codon adaptation index.\n",
      "     |          If multiple codons are equally preferred, a warning is issued\n",
      "     |          and one codon is chosen for use in the optimzed sequence.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __contains__(self, key, /)\n",
      "     |      True if the dictionary has the specified key, else False.\n",
      "     |  \n",
      "     |  __delitem__(self, key, /)\n",
      "     |      Delete self[key].\n",
      "     |  \n",
      "     |  __eq__(self, value, /)\n",
      "     |      Return self==value.\n",
      "     |  \n",
      "     |  __ge__(self, value, /)\n",
      "     |      Return self>=value.\n",
      "     |  \n",
      "     |  __getattribute__(self, name, /)\n",
      "     |      Return getattr(self, name).\n",
      "     |  \n",
      "     |  __getitem__(...)\n",
      "     |      x.__getitem__(y) <==> x[y]\n",
      "     |  \n",
      "     |  __gt__(self, value, /)\n",
      "     |      Return self>value.\n",
      "     |  \n",
      "     |  __ior__(self, value, /)\n",
      "     |      Return self|=value.\n",
      "     |  \n",
      "     |  __iter__(self, /)\n",
      "     |      Implement iter(self).\n",
      "     |  \n",
      "     |  __le__(self, value, /)\n",
      "     |      Return self<=value.\n",
      "     |  \n",
      "     |  __len__(self, /)\n",
      "     |      Return len(self).\n",
      "     |  \n",
      "     |  __lt__(self, value, /)\n",
      "     |      Return self<value.\n",
      "     |  \n",
      "     |  __ne__(self, value, /)\n",
      "     |      Return self!=value.\n",
      "     |  \n",
      "     |  __or__(self, value, /)\n",
      "     |      Return self|value.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __reversed__(self, /)\n",
      "     |      Return a reverse iterator over the dict keys.\n",
      "     |  \n",
      "     |  __ror__(self, value, /)\n",
      "     |      Return value|self.\n",
      "     |  \n",
      "     |  __setitem__(self, key, value, /)\n",
      "     |      Set self[key] to value.\n",
      "     |  \n",
      "     |  __sizeof__(...)\n",
      "     |      D.__sizeof__() -> size of D in memory, in bytes\n",
      "     |  \n",
      "     |  clear(...)\n",
      "     |      D.clear() -> None.  Remove all items from D.\n",
      "     |  \n",
      "     |  copy(...)\n",
      "     |      D.copy() -> a shallow copy of D\n",
      "     |  \n",
      "     |  get(self, key, default=None, /)\n",
      "     |      Return the value for key if key is in the dictionary, else default.\n",
      "     |  \n",
      "     |  items(...)\n",
      "     |      D.items() -> a set-like object providing a view on D's items\n",
      "     |  \n",
      "     |  keys(...)\n",
      "     |      D.keys() -> a set-like object providing a view on D's keys\n",
      "     |  \n",
      "     |  pop(...)\n",
      "     |      D.pop(k[,d]) -> v, remove specified key and return the corresponding value.\n",
      "     |      \n",
      "     |      If the key is not found, return the default if given; otherwise,\n",
      "     |      raise a KeyError.\n",
      "     |  \n",
      "     |  popitem(self, /)\n",
      "     |      Remove and return a (key, value) pair as a 2-tuple.\n",
      "     |      \n",
      "     |      Pairs are returned in LIFO (last-in, first-out) order.\n",
      "     |      Raises KeyError if the dict is empty.\n",
      "     |  \n",
      "     |  setdefault(self, key, default=None, /)\n",
      "     |      Insert key with a value of default if key is not in the dictionary.\n",
      "     |      \n",
      "     |      Return the value for key if key is in the dictionary, else default.\n",
      "     |  \n",
      "     |  update(...)\n",
      "     |      D.update([E, ]**F) -> None.  Update D from dict/iterable E and F.\n",
      "     |      If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k]\n",
      "     |      If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v\n",
      "     |      In either case, this is followed by: for k in F:  D[k] = F[k]\n",
      "     |  \n",
      "     |  values(...)\n",
      "     |      D.values() -> an object providing a view on D's values\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Class methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __class_getitem__(...) from builtins.type\n",
      "     |      See PEP 585\n",
      "     |  \n",
      "     |  fromkeys(iterable, value=None, /) from builtins.type\n",
      "     |      Create a new dictionary with keys from iterable and values set to value.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes inherited from builtins.dict:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "\n",
      "FUNCTIONS\n",
      "    GC123(seq)\n",
      "        Calculate G+C content: total, for first, second and third positions.\n",
      "        \n",
      "        Returns a tuple of four floats (percentages between 0 and 100) for the\n",
      "        entire sequence, and the three codon positions.  e.g.\n",
      "        \n",
      "        >>> from Bio.SeqUtils import GC123\n",
      "        >>> GC123(\"ACTGTN\")\n",
      "        (40.0, 50.0, 50.0, 0.0)\n",
      "        \n",
      "        Copes with mixed case sequences, but does NOT deal with ambiguous\n",
      "        nucleotides.\n",
      "    \n",
      "    GC_skew(seq, window=100)\n",
      "        Calculate GC skew (G-C)/(G+C) for multiple windows along the sequence.\n",
      "        \n",
      "        Returns a list of ratios (floats), controlled by the length of the sequence\n",
      "        and the size of the window.\n",
      "        \n",
      "        Returns 0 for windows without any G/C by handling zero division errors.\n",
      "        \n",
      "        Does NOT look at any ambiguous nucleotides.\n",
      "    \n",
      "    cos(x, /)\n",
      "        Return the cosine of x (measured in radians).\n",
      "    \n",
      "    exp(x, /)\n",
      "        Return e raised to the power of x.\n",
      "    \n",
      "    gc_fraction(seq, ambiguous='remove')\n",
      "        Calculate G+C percentage in seq (float between 0 and 1).\n",
      "        \n",
      "        Copes with mixed case sequences. Ambiguous Nucleotides in this context are\n",
      "        those different from ATCGSW (S is G or C, and W is A or T).\n",
      "        \n",
      "        If ambiguous equals \"remove\" (default), will only count GCS and will only\n",
      "        include ACTGSW when calculating the sequence length. Equivalent to removing\n",
      "        all characters in the set BDHKMNRVXY before calculating the GC content, as\n",
      "        each of these ambiguous nucleotides can either be in (A,T) or in (C,G).\n",
      "        \n",
      "        If ambiguous equals \"ignore\", it will treat only unambiguous nucleotides (GCS)\n",
      "        as counting towards the GC percentage, but will include all ambiguous and\n",
      "        unambiguous nucleotides when calculating the sequence length.\n",
      "        \n",
      "        If ambiguous equals \"weighted\", will use a \"mean\" value when counting the\n",
      "        ambiguous characters, for example, G and C will be counted as 1, N and X will\n",
      "        be counted as 0.5, D will be counted as 0.33 etc. See Bio.SeqUtils._gc_values\n",
      "        for a full list.\n",
      "        \n",
      "        Will raise a ValueError for any other value of the ambiguous parameter.\n",
      "        \n",
      "        \n",
      "        >>> from Bio.SeqUtils import gc_fraction\n",
      "        >>> seq = \"ACTG\"\n",
      "        >>> print(f\"GC content of {seq} : {gc_fraction(seq):.2f}\")\n",
      "        GC content of ACTG : 0.50\n",
      "        \n",
      "        S and W are ambiguous for the purposes of calculating the GC content.\n",
      "        \n",
      "        >>> seq = \"ACTGSSSS\"\n",
      "        >>> gc = gc_fraction(seq, \"remove\")\n",
      "        >>> print(f\"GC content of {seq} : {gc:.2f}\")\n",
      "        GC content of ACTGSSSS : 0.75\n",
      "        >>> gc = gc_fraction(seq, \"ignore\")\n",
      "        >>> print(f\"GC content of {seq} : {gc:.2f}\")\n",
      "        GC content of ACTGSSSS : 0.75\n",
      "        >>> gc = gc_fraction(seq, \"weighted\")\n",
      "        >>> print(f\"GC content with ambiguous counting: {gc:.2f}\")\n",
      "        GC content with ambiguous counting: 0.75\n",
      "        \n",
      "        Some examples with ambiguous nucleotides.\n",
      "        \n",
      "        >>> seq = \"ACTGN\"\n",
      "        >>> gc = gc_fraction(seq, \"ignore\")\n",
      "        >>> print(f\"GC content of {seq} : {gc:.2f}\")\n",
      "        GC content of ACTGN : 0.40\n",
      "        >>> gc = gc_fraction(seq, \"weighted\")\n",
      "        >>> print(f\"GC content with ambiguous counting: {gc:.2f}\")\n",
      "        GC content with ambiguous counting: 0.50\n",
      "        >>> gc = gc_fraction(seq, \"remove\")\n",
      "        >>> print(f\"GC content with ambiguous removing: {gc:.2f}\")\n",
      "        GC content with ambiguous removing: 0.50\n",
      "        \n",
      "        Ambiguous nucleotides are also removed from the length of the sequence.\n",
      "        \n",
      "        >>> seq = \"GDVV\"\n",
      "        >>> gc = gc_fraction(seq, \"ignore\")\n",
      "        >>> print(f\"GC content of {seq} : {gc:.2f}\")\n",
      "        GC content of GDVV : 0.25\n",
      "        >>> gc = gc_fraction(seq, \"weighted\")\n",
      "        >>> print(f\"GC content with ambiguous counting: {gc:.4f}\")\n",
      "        GC content with ambiguous counting: 0.6667\n",
      "        >>> gc = gc_fraction(seq, \"remove\")\n",
      "        >>> print(f\"GC content with ambiguous removing: {gc:.2f}\")\n",
      "        GC content with ambiguous removing: 1.00\n",
      "        \n",
      "        \n",
      "        Note that this will return zero for an empty sequence.\n",
      "    \n",
      "    log(...)\n",
      "        log(x, [base=math.e])\n",
      "        Return the logarithm of x to the given base.\n",
      "        \n",
      "        If the base not specified, returns the natural logarithm (base e) of x.\n",
      "    \n",
      "    molecular_weight(seq, seq_type='DNA', double_stranded=False, circular=False, monoisotopic=False)\n",
      "        Calculate the molecular mass of DNA, RNA or protein sequences as float.\n",
      "        \n",
      "        Only unambiguous letters are allowed. Nucleotide sequences are assumed to\n",
      "        have a 5' phosphate.\n",
      "        \n",
      "        Arguments:\n",
      "         - seq: string, Seq, or SeqRecord object.\n",
      "         - seq_type: The default is to assume DNA; override this with a string\n",
      "           \"DNA\", \"RNA\", or \"protein\".\n",
      "         - double_stranded: Calculate the mass for the double stranded molecule?\n",
      "         - circular: Is the molecule circular (has no ends)?\n",
      "         - monoisotopic: Use the monoisotopic mass tables?\n",
      "        \n",
      "        >>> print(\"%0.2f\" % molecular_weight(\"AGC\"))\n",
      "        949.61\n",
      "        >>> print(\"%0.2f\" % molecular_weight(Seq(\"AGC\")))\n",
      "        949.61\n",
      "        \n",
      "        However, it is better to be explicit - for example with strings:\n",
      "        \n",
      "        >>> print(\"%0.2f\" % molecular_weight(\"AGC\", \"DNA\"))\n",
      "        949.61\n",
      "        >>> print(\"%0.2f\" % molecular_weight(\"AGC\", \"RNA\"))\n",
      "        997.61\n",
      "        >>> print(\"%0.2f\" % molecular_weight(\"AGC\", \"protein\"))\n",
      "        249.29\n",
      "    \n",
      "    nt_search(seq, subseq)\n",
      "        Search for a DNA subseq in seq, return list of [subseq, positions].\n",
      "        \n",
      "        Use ambiguous values (like N = A or T or C or G, R = A or G etc.),\n",
      "        searches only on forward strand.\n",
      "    \n",
      "    seq1(seq, custom_map=None, undef_code='X')\n",
      "        Convert protein sequence from three-letter to one-letter code.\n",
      "        \n",
      "        The single required input argument 'seq' should be a protein sequence\n",
      "        using three-letter codes, either as a Python string or as a Seq or\n",
      "        MutableSeq object.\n",
      "        \n",
      "        This function returns the amino acid sequence as a string using the one\n",
      "        letter amino acid codes. Output follows the IUPAC standard (including\n",
      "        ambiguous characters \"B\" for \"Asx\", \"J\" for \"Xle\", \"X\" for \"Xaa\", \"U\" for\n",
      "        \"Sel\", and \"O\" for \"Pyl\") plus \"*\" for a terminator given the \"Ter\" code.\n",
      "        Any unknown character (including possible gap characters), is changed\n",
      "        into '-' by default.\n",
      "        \n",
      "        e.g.\n",
      "        \n",
      "        >>> from Bio.SeqUtils import seq1\n",
      "        >>> seq1(\"MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer\")\n",
      "        'MAIVMGRWKGAR*'\n",
      "        \n",
      "        The input is case insensitive, e.g.\n",
      "        \n",
      "        >>> from Bio.SeqUtils import seq1\n",
      "        >>> seq1(\"METalaIlEValMetGLYArgtRplysGlyAlaARGTer\")\n",
      "        'MAIVMGRWKGAR*'\n",
      "        \n",
      "        You can set a custom translation of the codon termination code using the\n",
      "        dictionary \"custom_map\" argument (defaulting to {'Ter': '*'}), e.g.\n",
      "        \n",
      "        >>> seq1(\"MetAlaIleValMetGlyArgTrpLysGlyAla***\", custom_map={\"***\": \"*\"})\n",
      "        'MAIVMGRWKGA*'\n",
      "        \n",
      "        You can also set a custom translation for non-amino acid characters, such\n",
      "        as '-', using the \"undef_code\" argument, e.g.\n",
      "        \n",
      "        >>> seq1(\"MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer\", undef_code='?')\n",
      "        'MAIVMGRWKGA??R*'\n",
      "        \n",
      "        If not given, \"undef_code\" defaults to \"X\", e.g.\n",
      "        \n",
      "        >>> seq1(\"MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer\")\n",
      "        'MAIVMGRWKGAXXR*'\n",
      "    \n",
      "    seq3(seq, custom_map=None, undef_code='Xaa')\n",
      "        Convert protein sequence from one-letter to three-letter code.\n",
      "        \n",
      "        The single required input argument 'seq' should be a protein sequence using\n",
      "        single letter codes, either as a Python string or as a Seq or MutableSeq\n",
      "        object.\n",
      "        \n",
      "        This function returns the amino acid sequence as a string using the three\n",
      "        letter amino acid codes. Output follows the IUPAC standard (including\n",
      "        ambiguous characters B for \"Asx\", J for \"Xle\" and X for \"Xaa\", and also U\n",
      "        for \"Sel\" and O for \"Pyl\") plus \"Ter\" for a terminator given as an\n",
      "        asterisk. Any unknown character (including possible gap characters),\n",
      "        is changed into 'Xaa' by default.\n",
      "        \n",
      "        e.g.\n",
      "        \n",
      "        >>> from Bio.SeqUtils import seq3\n",
      "        >>> seq3(\"MAIVMGRWKGAR*\")\n",
      "        'MetAlaIleValMetGlyArgTrpLysGlyAlaArgTer'\n",
      "        \n",
      "        You can set a custom translation of the codon termination code using the\n",
      "        dictionary \"custom_map\" argument (which defaults to {'*': 'Ter'}), e.g.\n",
      "        \n",
      "        >>> seq3(\"MAIVMGRWKGAR*\", custom_map={\"*\": \"***\"})\n",
      "        'MetAlaIleValMetGlyArgTrpLysGlyAlaArg***'\n",
      "        \n",
      "        You can also set a custom translation for non-amino acid characters, such\n",
      "        as '-', using the \"undef_code\" argument, e.g.\n",
      "        \n",
      "        >>> seq3(\"MAIVMGRWKGA--R*\", undef_code='---')\n",
      "        'MetAlaIleValMetGlyArgTrpLysGlyAla------ArgTer'\n",
      "        \n",
      "        If not given, \"undef_code\" defaults to \"Xaa\", e.g.\n",
      "        \n",
      "        >>> seq3(\"MAIVMGRWKGA--R*\")\n",
      "        'MetAlaIleValMetGlyArgTrpLysGlyAlaXaaXaaArgTer'\n",
      "        \n",
      "        This function was inspired by BioPerl's seq3.\n",
      "    \n",
      "    sin(x, /)\n",
      "        Return the sine of x (measured in radians).\n",
      "    \n",
      "    six_frame_translations(seq, genetic_code=1)\n",
      "        Return pretty string showing the 6 frame translations and GC content.\n",
      "        \n",
      "        Nice looking 6 frame translation with GC content - code from xbbtools\n",
      "        similar to DNA Striders six-frame translation\n",
      "        \n",
      "        >>> from Bio.SeqUtils import six_frame_translations\n",
      "        >>> print(six_frame_translations(\"AUGGCCAUUGUAAUGGGCCGCUGA\"))\n",
      "        GC_Frame: a:5 t:0 g:8 c:5\n",
      "        Sequence: auggccauug ... gggccgcuga, 24 nt, 54.17 %GC\n",
      "        <BLANKLINE>\n",
      "        <BLANKLINE>\n",
      "        1/1\n",
      "          G  H  C  N  G  P  L\n",
      "         W  P  L  *  W  A  A\n",
      "        M  A  I  V  M  G  R  *\n",
      "        auggccauuguaaugggccgcuga   54 %\n",
      "        uaccgguaacauuacccggcgacu\n",
      "        A  M  T  I  P  R  Q\n",
      "         H  G  N  Y  H  A  A  S\n",
      "          P  W  Q  L  P  G  S\n",
      "        <BLANKLINE>\n",
      "        <BLANKLINE>\n",
      "    \n",
      "    xGC_skew(seq, window=1000, zoom=100, r=300, px=100, py=100)\n",
      "        Calculate and plot normal and accumulated GC skew (GRAPHICS !!!).\n",
      "\n",
      "DATA\n",
      "    pi = 3.141592653589793\n",
      "    standard_dna_table = NCBICodonTableDNA(id=1, names=['Standard', 'SGC0'...\n",
      "\n",
      "FILE\n",
      "    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/SeqUtils/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import Bio.SeqUtils\n",
    "help(Bio.SeqUtils)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55.55555555555556\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "test_seq = Seq(\"ATGCAAGCGCTAGCGTAG\")\n",
    "gc_content = (test_seq.count(\"G\") + test_seq.count(\"C\")) / len(test_seq) * 100\n",
    "print(gc_content)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATGCGATAG', 'ATGCTAATAGCTAGTGAATGCGCGTAG']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_orfs(dna_sequence):\n",
    "    start_codon = 'ATG'\n",
    "    stop_codons = ['TAA', 'TAG', 'TGA']\n",
    "\n",
    "    def find_orfs_in_frame(sequence):\n",
    "        orfs = []\n",
    "        for i in range(0, len(sequence), 3):\n",
    "            if sequence[i:i+3] == start_codon:\n",
    "                for j in range(i+3, len(sequence), 3):\n",
    "                    if sequence[j:j+3] in stop_codons:\n",
    "                        orfs.append(sequence[i:j+3])\n",
    "                        break\n",
    "        return orfs\n",
    "\n",
    "    def reverse_complement(sequence):\n",
    "        complement = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "        return ''.join([complement[base] for base in sequence[::-1]])\n",
    "\n",
    "    orfs_found = []\n",
    "    for frame in range(3):\n",
    "        # Find ORFs in the original sequence\n",
    "        orfs_found.extend(find_orfs_in_frame(dna_sequence[frame:]))\n",
    "        # Find ORFs in the reverse complement sequence\n",
    "        orfs_found.extend(find_orfs_in_frame(reverse_complement(dna_sequence)[frame:]))\n",
    "    \n",
    "    return orfs_found\n",
    "\n",
    "# Test the function with a sample DNA sequence\n",
    "test_sequence = \"ATGCGATAGCGTATGCTAATAGCTAGTGAATGCGCGTAG\"\n",
    "orfs = find_orfs(test_sequence)\n",
    "orfs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ATG': [0, 12, 29], 'TAG': [6, 19, 23, 36], 'TGA': [26]}\n"
     ]
    }
   ],
   "source": [
    "def find_motifs(sequence, motifs):\n",
    "    \"\"\"\n",
    "    Find all occurrences of given motifs in a sequence.\n",
    "    \n",
    "    :param sequence: A string, the sequence in which to search for motifs.\n",
    "    :param motifs: A list of strings, the motifs to search for.\n",
    "    :return: A dictionary with each motif and a list of start positions where that motif is found.\n",
    "    \"\"\"\n",
    "    motif_positions = {motif: [] for motif in motifs}  # Initialize a dictionary to store positions\n",
    "    \n",
    "    for motif in motifs:\n",
    "        start = 0  # Start at the beginning of the sequence\n",
    "        while start < len(sequence):\n",
    "            pos = sequence.find(motif, start)  # Find the motif in the sequence from the current start position\n",
    "            if pos == -1:  # If the motif is not found, break out of the loop\n",
    "                break\n",
    "            motif_positions[motif].append(pos)  # Add the found position to the list for this motif\n",
    "            start = pos + 1  # Move the start position past the last found position\n",
    "            \n",
    "    return motif_positions\n",
    "\n",
    "# Example usage:\n",
    "sequence = \"ATGCGATAGCGTATGCTAATAGCTAGTGAATGCGCGTAG\"\n",
    "motifs = [\"ATG\", \"TAG\", \"TGA\"]\n",
    "found_motifs = find_motifs(sequence, motifs)\n",
    "print(found_motifs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Motif 'ATGC' found at positions: [0, 12, 29]\n"
     ]
    }
   ],
   "source": [
    "def find_motifs(sequence, motif):\n",
    "    motif_locations = []\n",
    "    motif_length = len(motif)\n",
    "    for i in range(len(sequence) - motif_length + 1):\n",
    "        if sequence[i:i+motif_length] == motif:\n",
    "            motif_locations.append(i)\n",
    "    return motif_locations\n",
    "\n",
    "# Example usage:\n",
    "sequence = \"ATGCGATAGCGTATGCTAATAGCTAGTGAATGCGCGTAG\"\n",
    "motif = \"ATGC\"\n",
    "motif_locations = find_motifs(sequence, motif)\n",
    "print(f\"Motif '{motif}' found at positions: {motif_locations}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11.5 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
